tools/perf/ResultsComparer/src/ResultsComparer/Core/ResultsComparerProvider.cs

```diff
@@ -0,0 +1,64 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ResultsComparerProvider.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+
+namespace ResultsComparer.Core
+{
+    /// <summary>
+    /// The default <see cref="IResultsComparerProvider"/>.
+    /// </summary>
+    public class ResultsComparerProvider : IResultsComparerProvider
+    {
+        Dictionary<string, IResultsComparer> comparers = new();
+        // the list is used to maintain the order of insertion of comparers
+        // because we want to test the comparers by order of priority
+        // when auto-detecting suitable comparers
+        List<IResultsComparer> comparersList = new();
+
+        /// <summary>
+        /// Registers the specified <see cref="IResultsComparerer"/> with
+        /// the specified <paramref name="id"/>.
+        /// </summary>
+        /// <param name="id">The ID to associate with the comparer for later retrieval.</param>
+        /// <param name="comparer">The comparer to register.</param>
+        public void RegisterComparer(string id, IResultsComparer comparer)
+        {
+            if (!comparers.TryAdd(id, comparer))
+            {
+                throw new Exception($"A comparer with id '{comparer}' has already been registered.");
+            }
+
+            comparersList.Add(comparer);
+        }
+
+        /// <inheritdoc/>
+        public IResultsComparer GetById(string comparerId)
+        {
+            if (comparers.TryGetValue(comparerId, out var comparer))
+            {
+                return comparer;
+            }
+
+            throw new Exception($"Unknown comparer '{comparerId}'.");

```

Comment:
Is throwing an exception when an `IResultsComparer` matching the specified `comparerId` better than returning null?

Reply:
My reasoning was this if you return null, then you always have to check that the return value is not null before using it. Since the code executes everything is already in a try-catch block, then you don't need to introduce any additional blocks to handle situations like this. I also thought it's easier to find out what's wrong if the app fail with a clear exception message than if it fails due to some random null-pointer exception that can appear at a different site from where this method was called.

It does feel too much to throw an exception just because an id wasn't found. But based on the intended use case, you should not be trying to fetch IDs that don't exist, this is not a general purpose key-value store. If you're fetching an ID that does not exist, it could be an indication of a typo or some bug somewhere in the code, so I think an exception is appropriate.

What do you think?

Reply:
Makes sense